src/hotspot/os/linux/cgroupV1Subsystem_linux.cpp

```diff
@@ -89,80 +89,84 @@ void CgroupV1MemoryController::set_subsystem_path(char *cgroup_path) {
   host_total_memsw = os::Linux::host_swap() + os::Linux::physical_memory();
-  if (memswlimit >= host_total_memsw) {
+  if ((julong)memswlimit >= host_total_memsw) {
     log_trace(os, container)("Non-Hierarchical Memory and Swap Limit is: Unlimited");
     CgroupV1MemoryController* mem_controller = reinterpret_cast<CgroupV1MemoryController*>(_memory->controller());
     if (mem_controller->is_hierarchical()) {
       const char* matchline = "hierarchical_memsw_limit";
-      GET_CONTAINER_INFO_LINE(julong, _memory->controller(), "/memory.stat", matchline,
-                             "Hierarchical Memory and Swap Limit is : " JULONG_FORMAT, JULONG_FORMAT, hier_memswlimit)
-      if (hier_memswlimit >= host_total_memsw) {
+      GET_CONTAINER_INFO_LINE(jlong, _memory->controller(), "/memory.stat", matchline,
+                             "Hierarchical Memory and Swap Limit is : " JLONG_FORMAT, JLONG_FORMAT, hier_memswlimit)
+      assert(hier_memswlimit > 0, "invariant");
+      if ((julong)hier_memswlimit >= host_total_memsw) {
         log_trace(os, container)("Hierarchical Memory and Swap Limit is: Unlimited");
       } else {
         jlong swappiness = read_mem_swappiness();
         if (swappiness == 0) {
             const char* matchmemline = "hierarchical_memory_limit";
-            GET_CONTAINER_INFO_LINE(julong, _memory->controller(), "/memory.stat", matchmemline,
-                             "Hierarchical Memory Limit is : " JULONG_FORMAT, JULONG_FORMAT, hier_memlimit)
-            log_trace(os, container)("Memory and Swap Limit has been reset to " JULONG_FORMAT " because swappiness is 0", hier_memlimit);
-            return (jlong)hier_memlimit;
+            GET_CONTAINER_INFO_LINE(jlong, _memory->controller(), "/memory.stat", matchmemline,
+                             "Hierarchical Memory Limit is : " JLONG_FORMAT, JLONG_FORMAT, hier_memlimit)
+            log_trace(os, container)("Memory and Swap Limit has been reset to " JLONG_FORMAT " because swappiness is 0", hier_memlimit);
+            return hier_memlimit;
         }
-        return (jlong)hier_memswlimit;
+        return hier_memswlimit;
       }
     }
     return (jlong)-1;
   } else {
     jlong swappiness = read_mem_swappiness();
     if (swappiness == 0) {
       jlong memlimit = read_memory_limit_in_bytes();
-      log_trace(os, container)("Memory and Swap Limit has been reset to " JULONG_FORMAT " because swappiness is 0", memlimit);
+      log_trace(os, container)("Memory and Swap Limit has been reset to " JLONG_FORMAT " because swappiness is 0", memlimit);
       return memlimit;
     }
-    return (jlong)memswlimit;
+    return memswlimit;
   }
 }
 
 jlong CgroupV1Subsystem::read_mem_swappiness() {
-  GET_CONTAINER_INFO(julong, _memory->controller(), "/memory.swappiness",
-                     "Swappiness is: " JULONG_FORMAT, JULONG_FORMAT, swappiness);
+  GET_CONTAINER_INFO(jlong, _memory->controller(), "/memory.swappiness",
```

Comment:
Why is this changed from julong to jlong? The only caller of this (private) function is to check the returned value against zero. It seems like this function should return julong instead to avoid all the casting and possible overflow/underflow.

Reply:
> Why is this changed from julong to jlong?

Because `GET_CONTAINER_INFO` might log lines such as these (with `-Xlog:os+container=trace`):
```
Swappiness is: 18446744073709551614
```
... on systems where for example `swapaccount=0` is being set on the kernel comandline. That is actually `-2`, i.e. `OS_CONTAINER_ERROR`. Which is just very confusing.

That file, `memory.swappiness`, actually contains small numbers (if it exists). So there is really no need to read this in as `julong`. With this change it would print on `swapaccount=0` systems:

```
Swappiness is: -2
```

There is really no need to read as `julong`.

Reply:
Is the motivation of this PR that `GET_CONTAINER_INFO` can't correctly print the error log for unsigned types?

```
    log_trace(os, container)(logstring, (return_type) OSCONTAINER_ERROR); \
```

If so, I think it's better to fix the logging code so that it can handle unsigned types, and not change all values to signed types. Fixing logging is a lot easier than verifying that your type casting and range clipping is correct.

Reply:
Yes.